packages/components/src/lib/controllers/settings-controller.js

```diff
@@ -0,0 +1,264 @@
+        // to redraw themselves when settings are reset or reloaded
+        // it might be better if all settings were made into
+        // state variables but for now state is monitored at the domain level
+        uiResetCounter: 0,
+        featureResetCounter: 0,
+        resourceResetCounter: 0
+      },
+      mutations: {
+        incrementUiResetCounter (state) {
+          state.uiResetCounter += 1
+        },
+
+        incrementFeatureResetCounter (state) {
+          state.featureResetCounter += 1
+        },
+
+        incrementResourceResetCounter (state) {
+          state.resourceResetCounter += 1
+        }
+      }
+    })
+  }
+
+  /**
+   * Loads an application wide configuration file in a JSON format.
+   *
+   * @returns {Promise<object>} - A promise that is resolved with an app config response.
+   */
+  async requestAppOptions () {
+    const configUrl = `${this._configServiceUrl}?clientId=${encodeURIComponent(this._clientId)}&appName=${encodeURIComponent(this._appName)}` +
+      `&appVersion=${encodeURIComponent(this._appVersion)}&buildBranch=${encodeURIComponent(this._branch)}` +
+      `&buildNumber=${encodeURIComponent(this._buildNumber)}`
+    const request = new Request(configUrl)
+    return fetch(request)
+  }
+
+  /**
+   * initialize the options using the supplied storage adapter class
+   *
+   * @param {Function<StorageAdapter>} StorageAdapter the adapter class to instantiate
+   * @param {object} authData optional authentication data if the adapter is one that requires it
+   * @returns Promise[] an array of promises to load the options data from the adapter
+   */
+  initOptions (StorageAdapter = this._storageAdapter, authData = null) {
+    this._featureOptions = new Options(this._featureOptionsDefaults, new StorageAdapter(this._featureOptionsDefaults.domain, authData))
+    this._resourceOptions = new Options(this._resourceOptionsDefaults, new StorageAdapter(this._resourceOptionsDefaults.domain, authData))
+    this._uiOptions = new Options(this._uiOptionsDefaults, new StorageAdapter(this._uiOptionsDefaults.domain, authData))
+    return [this._featureOptions.load(), this._resourceOptions.load(), this._uiOptions.load()]
+  }
+
+  initNonConfigurableOptions () {
+    this._lookupResourceOptions = new Options(this._resourceOptionsDefaults, new this._storageAdapter(this._resourceOptionsDefaults.domain))
+    this._siteOptions = [] // eslint-disable-line prefer-const
+    for (const site of this._siteOptionsDefaults) {
+      for (const domain of site.options) {
+        const siteOpts = new Options(domain, new this._storageAdapter(domain.domain)) // eslint-disable-line new-cap
+        this._siteOptions.push({ uriMatch: site.uriMatch, resourceOptions: siteOpts })
+      }
+    }
+  }
+
+  getFeatureOptions () {
+    return this._featureOptions
+  }
+
+  getResourceOptions () {
+    return this._resourceOptions
+  }
+
+  getUiOptions () {
```

Comment:
The options that start with `_` are private, they are hidden within the Settings Controller. I think it's good to keep them isolated because it will give us freedom to change them in future (at least theoretically). `getABCOptions()` methods are the public API of the Settings Controller. I think it's better to have them as methods, not as properties, because of the following advantages:
1. It allows to hide internal implementation. As long as the `get...()` method returns the options object, it does not matter where it takes this information internally within the Settings Controller object. It can for example, take settings that are kept in two different properties and combine them into one structure.
2. Methods can accept options. For example, clients could use `getABCOptions()` to get an editable version of the options or call `getABCOptions({ readOnly: true })` that will return a copy of the options object.
3. It can transform data. For example, if we change the options format internally, we could make the getter method to transform it the way that it would be compatible with the old format. This will allow to keep compatibility.
4. If accessors are methods, it is easier to track access calls. Might be useful for debugging. We could add logger statements into the methods and know what part of the app accesses options and in what order calls are made.
Those are theoretical benefits. We do not use them now, mostly. So I'm not sure if we should keep them as methods for the future uses or turn them to preperties because that will be a little simpler. @balmas, @irina060981, what do you think?

Reply:
In theory it sounds really good, but in our case they are not really private.
You pass them as a direct link through the additional step.
And they could be easily changed outside.

But if you want to place here the template for future, I am not against of course.

Reply:
I would keep them as functions for the future, if @balmas is not against it 🙂. One real benefit of having them as functions that I've used was tracking: log statements put inside those methods helped me a lot to figure out a tricky option initialization and access sequence.

Reply:
Well with javascript, the whole concept of public vs private is really a matter of coding practice and not enforced by anything.

I do appreciate the questions and close look at the code from @irina060981 but I think I'm fine with this as it is.